### DISCUSSION TITLE
Does emissive work for UNLIT material.

### DISCUSSION THREAD

**Discussion by JackKaiXing** [google/filament#4791] [Q&A] [answered]
Hello,

I tried with three glTF models.

| | extension | emissiveFactor | hasEmissiveOnRenderResult
|-----|-----|-----|-----|
glTF-A | KHR_materials_unlit | emissiveFactor: {31.415926, 0.0, 0.0} | No
glTF-B | ..... | emissiveFactor: {31.415926, 0.0, 0.0} | Yes
glTF-C | KHR_materials_unlit | emissiveFactor: {31.415926 * 10000.0, 0.0, 0.0} | `Yes`

Any ideas?

**Comment by romainguy** [comment, marked as answer]
Yes it should work.

**Reply by JackKaiXing** [reply]
@romainguy 

But why should the emissiveFactor be modified to a huge value `{31.415926 * 10000.0, 0.0, 0.0}` to make emissive visible for unlit color? As the above glTF-C implies.

**Reply by romainguy** [reply]
You shouldn't have to. Is the emissive otherwise invisible?

**Reply by JackKaiXing** [reply]
A low value of emissiveFactor for `lit glTF ` makes the effects `visible`.
A low value of emissiveFactor for `unlit glTF ` makes the effects `invisible`.
a high value of  emissiveFactor for `unlit glTF ` makes the effects `visible`.

**Reply by JackKaiXing** [reply]
And I will upload the three glTF files for your test, in case you need.

**Reply by romainguy** [reply]
I checked and the logic is the same between unlit and lit materials. Here's an example of emissive red in lit and unlit materials:
<img width="1616" alt="Screen Shot 2021-11-01 at 8 39 23 AM" src="https://user-images.githubusercontent.com/869684/139698771-d94309ee-f958-48a7-b52e-5fe0a0d808ce.png">
<img width="1616" alt="Screen Shot 2021-11-01 at 8 39 26 AM" src="https://user-images.githubusercontent.com/869684/139698787-72e0549e-623f-475b-ac6b-f5aaa8375361.png">

**Reply by JackKaiXing** [reply]
I tried the following glTF file. https://github.com/JackKaiXing/glTF

The only difference between those two boxes is that the right one has "KHR_materials_unlit". But the right has no emissive effects.

![screenshot](https://user-images.githubusercontent.com/21079351/139782062-c6fb4eac-c078-4df2-915d-cc3ba83b32d4.png)

**Reply by romainguy** [reply]
Thanks, I'll take a look tomorrow.

**Reply by JackKaiXing** [reply]
For `ubershader` unlit material, this line would not be called.

https://github.com/google/filament/blob/78b29fe967dcea9bb4aff7c18386e25a6aa5b085/libs/gltfio/materials/ubershader.mat.in#L128

So for this line, I believe, the emissive.rgb is zero.
https://github.com/google/filament/blob/78b29fe967dcea9bb4aff7c18386e25a6aa5b085/shaders/src/shading_unlit.fs#L5

**Reply by JackKaiXing** [reply]
I believe, it is the same for GENERATE_SHADERS, since the value of emissiveFactor is not assigned to material.emissive.

**Reply by JackKaiXing** [reply]
Another problem should be about `emissive.w`.

It is initialized as 1.0.
https://github.com/google/filament/blob/78b29fe967dcea9bb4aff7c18386e25a6aa5b085/shaders/src/material_inputs.fs#L105

It is changed to 0.0 for lit material, but remains for 1.0 for unlit material.
https://github.com/google/filament/blob/78b29fe967dcea9bb4aff7c18386e25a6aa5b085/libs/gltfio/materials/ubershader.mat.in#L128

which leaves for the difference of `attenuation` value for lit and unlit material.
https://github.com/google/filament/blob/78b29fe967dcea9bb4aff7c18386e25a6aa5b085/shaders/src/shading_unlit.fs#L4

**Reply by JackKaiXing** [reply]
After hard-coded for unlit material, it seems work?
```
material.emissive.r = 3.1415926; // the same as lit material
material.emissive.w = 0.0;
```
![screenshot](https://user-images.githubusercontent.com/21079351/139833326-7d38efa7-3321-4280-a917-a1b2cb1bc583.png)

Any idea?

**Reply by romainguy** [reply]
I haven't looked at it yet but checking w was going to be my first step. The more important issue is that Filament technically follows the glTF spec that states:

> When present, the extension indicates that a material should be unlit and use available baseColor values, alpha values, and vertex colors while ignoring all properties of the default PBR model related to lighting or color.

Under the specification, `KHR_materials_unlit` does not support emissive.

**Reply by romainguy** [reply]
Also note that according to the spec, emissive values cannot be > 1.0 (we support them though).

**Reply by romainguy** [reply]
Last but not least, Filament follows the behavior of the reference glTF viewer (https://github.khronos.org/glTF-Sample-Viewer-Release/) as well as other viewers I've tried. So unfortunately you'll have to use "lit" to use emissive.
